### PR TITLE
support basic or digest auth for OTA restore

### PR DESCRIPTION
**Pull request by nickpell**
fixes http://manage.dimagi.com/default.asp?177414

### CODE REVIEW

**Comment by nickpell** [comment]
@dannyroberts @orangejenny 

Ota restore works locally.  Since the issue was only observed on the India server, after merging I'll deploy to India and confirm with @phillipm that the problem is fixed.

**Comment by nickpell** [comment]
:white_check_mark:

**Comment by orangejenny** [comment]
looks fine to me, would rather @dannyroberts merge

File: corehq/apps/domain/decorators.py [hunk 1]
```diff
@@ -178,6 +179,16 @@ def login_or_basic_ex(allow_cc_users=False):
 login_or_basic = login_or_basic_ex()
 
 
+def login_or_digest_or_basic(fn):
+    @wraps(fn)
+    def _inner(request, *args, **kwargs):
+        return {
+            'basic': login_or_basic_ex(allow_cc_users=True),
+            'digest': login_or_digest_ex(allow_cc_users=True),
+        }[determine_authtype(request)](fn)(request, *args, **kwargs)
```
Comment (dannyroberts) [inline corehq/apps/domain/decorators.py]: This is a bit of a nitpick, but if `determine_authtype` and `guess_phone_type_from_user_agent` are going to be used more globally like this, I'd prefer not having them stay in `corehq.apps.receiverwrapper.util`.

Additionally, `determine_authtype` pulls from the `authtype` GET param, so `?authtype=digest` will override the guessing, and `?authtype=yourmom` will make the URL 500 with a `KeyError`. I'm not sure either of those is specifically desired, so it seems wrong to have those behaviors get roped in as well.

**Comment by czue** [comment]
@nickpell when you say "ota restore works locally" do you mean that you tested from a phone?

**Comment by nickpell** [comment]
@czue Yes, that's what I meant to say (using my computer as the server).

File: corehq/apps/domain/decorators.py (hunk 1 above)
Comment (nickpell) [inline corehq/apps/domain/decorators.py]: good point about ```authtype``` - forgot to deal with that part of the receiver logic.

do you have a recommendation for an alternative module?

Comment (nickpell) [inline corehq/apps/domain/decorators.py]: fixed the ```authtype``` issue - let me know if you have a recommendation for module location

Comment (dannyroberts) [inline corehq/apps/domain/decorators.py]: Anywhere that's a more or less "global" file (i.e. a file that's imported indescriminantly to other apps throughout the project). I think even just putting it in the same file (domain/decorators.py) is fine

**Comment by czue** [comment]
:+1: from me but i'll also defer to @dannyroberts on merge.

when you tested from a phone were able to verify it used basic auth? or just that it went through? might be good to do some quick tests in curl to confirm it supports both.

**Comment by dannyroberts** [comment]
@nickpell I'll offline last steps with you. Seems like this is probably good to go, but want to check with you about QA

**Comment by dannyroberts** [comment]
Just tested with 
```bash
curl --digest -u $username@$domain.commcarehq.org https://staging.commcarehq.org/a/$domain/phone/restore/?version=2.0 -v --header 'User-Agent: Nokia'
```

**Pull request merged**